sdks/python/apache_beam/pipeline.py

```diff
@@ -671,7 +671,11 @@ def apply(
         # If the user wants the old implementation of always generated
         # PCollection output ids, then set the tag to None first, then count up
         # from 1.
-        tag = len(current.outputs) if None in current.outputs else None
+        base = tag
+        counter = 0
+        while tag in current.outputs:
+          counter += 1
+          tag = '%s_%d' % (base, counter)
         current.add_output(result, tag)
```

Comment:
Should this be under the experiment flag instead, so that we don't inadvertently break anyone?

Reply:
I am relatively confident in this change, as it preserves the essential characteristic (that output names are unique) and defaults to the same thing for all single-output transforms. However, I have added the opt-out you had originally with a note just in case.